File: sdk-typescript/src/governance/policy-template-v1.ts
import { createHash } from 'node:crypto';
import { canonicalJsonForHash } from '../policy/tool-pinning.js';
import { deriveDeviceKeyId, verifyDeviceSig, type DeviceSigner } from '../proxy/device-identity.js';
import { compareCodePoints } from './strict-canonical.js';

export const POLICY_TEMPLATE_V1_SCHEMA = 'obsvr-policy-template-v1' as const;
export const RENDERED_POLICY_V1_SCHEMA = 'obsvr-rendered-policy-v1' as const;
export const RENDERED_POLICY_ENVELOPE_V1_SCHEMA = 'obsvr-rendered-policy-envelope-v1' as const;
const HASH_RE = /^[0-9a-f]{64}$/;
type Json = null | boolean | number | string | Json[] | { [key: string]: Json };
export interface TemplateParameterV1 { name: string; type: 'string' | 'integer' | 'boolean' | 'enum'; enum_values?: string[]; }
export interface PolicyTemplateV1Input { template_id: string; version: string; parameters: TemplateParameterV1[]; artifact: Json; }

export class PolicyTemplateV1ValidationError extends Error { constructor(message: string) { super(message); this.name = 'PolicyTemplateV1ValidationError'; } }
function fail(message: string): never { throw new PolicyTemplateV1ValidationError(message); }
function text(value: unknown, field: string, max = 2048): string { if (typeof value !== 'string' || !value.trim() || Buffer.byteLength(value.trim(), 'utf8') > max) fail(`${field} must be nonblank and at most ${max} UTF-8 bytes`); return value.trim(); }
function hash(value: unknown, field: string): string { if (typeof value !== 'string' || !HASH_RE.test(value)) fail(`${field} must be a lowercase SHA-256 hash`); return value; }
function digest(domain: string, value: unknown): string { return createHash('sha256').update(`${domain}\0${canonicalJsonForHash(value)}`, 'utf8').digest('hex'); }
function validateJson(value: unknown, depth = 0, count = { value: 0 }): asserts value is Json {
  count.value += 1; if (count.value > 4096 || depth > 16) fail('template artifact exceeds structural bounds');
  if (value === null || typeof value === 'boolean' || typeof value === 'string') { if (typeof value === 'string') text(value, 'template string'); return; }
  if (typeof value === 'number') { if (!Number.isSafeInteger(value)) fail('template numbers must be safe integers'); return; }
  if (Array.isArray(value)) { if (value.length > 256) fail('template array exceeds 256 items'); value.forEach((item) => validateJson(item, depth + 1, count)); return; }
  if (!value || typeof value !== 'object') fail('template contains unsupported JSON value');
  const record = value as Record<string, unknown>; if (Object.keys(record).length > 256) fail('template object exceeds 256 fields');
  for (const [key, item] of Object.entries(record)) { text(key, 'template key', 256); validateJson(item, depth + 1, count); }
}

export function buildPolicyTemplateV1(input: PolicyTemplateV1Input) {
  if (!input || typeof input !== 'object' || Array.isArray(input)) fail('template must be an object');
  const raw = input as unknown as Record<string, unknown>; const unknown = Object.keys(raw).filter((key) => !['schema', 'template_id', 'version', 'parameters', 'artifact'].includes(key)).sort(compareCodePoints); if (unknown.length) fail(`template contains unsupported field: ${unknown[0]}`);
  if (raw.schema !== undefined && raw.schema !== POLICY_TEMPLATE_V1_SCHEMA) fail('template schema is invalid');
  if (!Array.isArray(raw.parameters) || raw.parameters.length > 128) fail('parameters must contain at most 128 items');
  const parameters = raw.parameters.map((candidate, index) => {
    if (!candidate || typeof candidate !== 'object' || Array.isArray(candidate)) fail(`parameters[${index}] must be an object`);
    const p = candidate as Record<string, unknown>; const extra = Object.keys(p).filter((key) => !['name', 'type', 'enum_values'].includes(key)); if (extra.length) fail(`parameters[${index}] contains unsupported field: ${extra[0]}`);
    if (!['string', 'integer', 'boolean', 'enum'].includes(String(p.type))) fail(`parameters[${index}].type is invalid`);
    const result: TemplateParameterV1 = { name: text(p.name, `parameters[${index}].name`, 256), type: p.type as TemplateParameterV1['type'] };
    if (result.type === 'enum') { if (!Array.isArray(p.enum_values) || p.enum_values.length === 0 || p.enum_values.length > 128) fail(`parameters[${index}].enum_values is required`); result.enum_values = [...new Set(p.enum_values.map((v, i) => text(v, `enum_values[${i}]`, 256)))].sort(compareCodePoints); }
    else if (p.enum_values !== undefined) fail(`parameters[${index}].enum_values is only valid for enum`);
    return result;
  }).sort((a, b) => compareCodePoints(a.name, b.name));
  if (new Set(parameters.map((item) => item.name)).size !== parameters.length) fail('parameter names must be unique');
  validateJson(raw.artifact);
  return { schema: POLICY_TEMPLATE_V1_SCHEMA, template_id: text(raw.template_id, 'template_id', 256), version: text(raw.version, 'version', 256), parameters, artifact: raw.artifact as Json };
}

export function policyTemplateV1Hash(input: PolicyTemplateV1Input): string { return digest('obsvr-policy-template/1', buildPolicyTemplateV1(input)); }

function validateParam(parameter: TemplateParameterV1, value: unknown): Json {
  if (parameter.type === 'string' && typeof value === 'string') return text(value, parameter.name);
  if (parameter.type === 'integer' && typeof value === 'number' && Number.isSafeInteger(value)) return value;
  if (parameter.type === 'boolean' && typeof value === 'boolean') return value;
  if (parameter.type === 'enum' && typeof value === 'string' && parameter.enum_values?.includes(value)) return value;
  return fail(`parameter ${parameter.name} does not match type ${parameter.type}`);
}
function renderNode(value: Json, params: Record<string, Json>): Json {
  if (Array.isArray(value)) return value.map((item) => renderNode(item, params));
  if (value && typeof value === 'object') {
    const keys = Object.keys(value);
    if (keys.length === 1 && keys[0] === '$obsvr_param') { const name = (value as Record<string, Json>).$obsvr_param; if (typeof name !== 'string' || !(name in params)) fail('template references an undeclared parameter'); return params[name]; }
    return Object.fromEntries(Object.entries(value).map(([key, item]) => [key, renderNode(item, params)]));
  }
  return value;
}

export function renderPolicyTemplateV1(templateInput: PolicyTemplateV1Input, supplied: Record<string, unknown>, approvalHash: string, activationHash: string) {
  const template = buildPolicyTemplateV1(templateInput); if (!supplied || typeof supplied !== 'object' || Array.isArray(supplied)) fail('supplied parameters must be an object');
  const expected = new Set(template.parameters.map((item) => item.name)); const extras = Object.keys(supplied).filter((key) => !expected.has(key)); if (extras.length) fail(`undeclared parameter supplied: ${extras.sort(compareCodePoints)[0]}`);
  const params: Record<string, Json> = {}; for (const parameter of template.parameters) { if (!(parameter.name in supplied)) fail(`missing parameter: ${parameter.name}`); params[parameter.name] = validateParam(parameter, supplied[parameter.name]); }
  const rendered = renderNode(template.artifact, params); validateJson(rendered);
  return { schema: RENDERED_POLICY_V1_SCHEMA, template_id: template.template_id, template_version: template.version, template_hash: policyTemplateV1Hash(template), parameters: Object.fromEntries(Object.entries(params).sort(([a], [b]) => compareCodePoints(a, b))), parameters_hash: digest('obsvr-policy-template-parameters/1', params), rendered_artifact: rendered, artifact_hash: digest('obsvr-rendered-policy-artifact/1', rendered), approval_hash: hash(approvalHash, 'approval_hash'), activation_hash: hash(activationHash, 'activation_hash') };
}

export function signRenderedPolicyV1(rendered: ReturnType<typeof renderPolicyTemplateV1>, signer: DeviceSigner) { const bodyHash = digest('obsvr-rendered-policy/1', rendered); const payload = `obsvr-rendered-policy-signature/1\0${bodyHash}`; return { schema: RENDERED_POLICY_ENVELOPE_V1_SCHEMA, body: rendered, body_hash: bodyHash, key_id: signer.keyId, signature: signer.signPayload(payload) }; }
export function verifyRenderedPolicyV1(envelope: ReturnType<typeof signRenderedPolicyV1>, rawPublicKey: Buffer): boolean { try { const bodyHash = digest('obsvr-rendered-policy/1', envelope.body); return envelope.schema === RENDERED_POLICY_ENVELOPE_V1_SCHEMA && envelope.body_hash === bodyHash && envelope.key_id === deriveDeviceKeyId(rawPublicKey) && verifyDeviceSig(rawPublicKey, envelope.key_id, `obsvr-rendered-policy-signature/1\0${bodyHash}`, envelope.signature); } catch { return false; } }

File: sdk-python/obsvr/policy_template_v1.py
"""Typed, signed policy templates with exact rendered provenance."""

from __future__ import annotations

import hashlib
from typing import Any, Dict

from .device_identity import DeviceSigner, derive_device_key_id, verify_device_sig
from .strict_canonical import code_point_key
from .tool_pinning import _canonical_json_for_hash

POLICY_TEMPLATE_V1_SCHEMA = "obsvr-policy-template-v1"
RENDERED_POLICY_V1_SCHEMA = "obsvr-rendered-policy-v1"
RENDERED_POLICY_ENVELOPE_V1_SCHEMA = "obsvr-rendered-policy-envelope-v1"
_HEX = frozenset("0123456789abcdef")


class PolicyTemplateV1ValidationError(ValueError):
    pass


def _fail(message: str) -> None:
    raise PolicyTemplateV1ValidationError(message)


def _text(value: Any, field: str, maximum: int = 2048) -> str:
    if not isinstance(value, str) or not value.strip() or len(value.strip().encode()) > maximum:
        _fail(f"{field} must be nonblank and at most {maximum} UTF-8 bytes")
    return value.strip()


def _hash(value: Any, field: str) -> str:
    if not isinstance(value, str) or len(value) != 64 or any(c not in _HEX for c in value):
        _fail(f"{field} must be a lowercase SHA-256 hash")
    return value


def _digest(domain: str, value: Any) -> str:
    return hashlib.sha256(f"{domain}\0{_canonical_json_for_hash(value)}".encode()).hexdigest()


def _validate_json(value: Any, depth: int = 0, counter: list[int] | None = None) -> None:
    counter = counter or [0]
    counter[0] += 1
    if counter[0] > 4096 or depth > 16:
        _fail("template artifact exceeds structural bounds")
    if value is None or isinstance(value, bool):
        return
    if isinstance(value, str):
        _text(value, "template string")
        return
    if isinstance(value, int) and not isinstance(value, bool) and abs(value) <= 9_007_199_254_740_991:
        return
    if isinstance(value, list):
        if len(value) > 256:
            _fail("template array exceeds 256 items")
        for item in value:
            _validate_json(item, depth + 1, counter)
        return
    if isinstance(value, dict):
        if len(value) > 256:
            _fail("template object exceeds 256 fields")
        for key, item in value.items():
            _text(key, "template key", 256)
            _validate_json(item, depth + 1, counter)
        return
    _fail("template contains unsupported JSON value")


def build_policy_template_v1(input_value: Dict[str, Any]) -> Dict[str, Any]:
    if not isinstance(input_value, dict):
        _fail("template must be an object")
    unknown = sorted(set(input_value) - {"schema", "template_id", "version", "parameters", "artifact"}, key=code_point_key)
    if unknown:
        _fail(f"template contains unsupported field: {unknown[0]}")
    if "schema" in input_value and input_value["schema"] != POLICY_TEMPLATE_V1_SCHEMA:
        _fail("template schema is invalid")
    raw_parameters = input_value.get("parameters")
    if not isinstance(raw_parameters, list) or len(raw_parameters) > 128:
        _fail("parameters must contain at most 128 items")
    parameters = []
    for index, item in enumerate(raw_parameters):
        if not isinstance(item, dict):
            _fail(f"parameters[{index}] must be an object")
        extra = set(item) - {"name", "type", "enum_values"}
        if extra:
            _fail(f"parameters[{index}] contains unsupported field: {sorted(extra)[0]}")
        if item.get("type") not in {"string", "integer", "boolean", "enum"}:
            _fail(f"parameters[{index}].type is invalid")
        result = {"name": _text(item.get("name"), f"parameters[{index}].name", 256), "type": item["type"]}
        if result["type"] == "enum":
            values = item.get("enum_values")
            if not isinstance(values, list) or not 1 <= len(values) <= 128:
                _fail(f"parameters[{index}].enum_values is required")
            result["enum_values"] = sorted({_text(v, "enum value", 256) for v in values}, key=code_point_key)
        elif "enum_values" in item:
            _fail(f"parameters[{index}].enum_values is only valid for enum")
        parameters.append(result)
    parameters.sort(key=lambda item: code_point_key(item["name"]))
    if len({item["name"] for item in parameters}) != len(parameters):
        _fail("parameter names must be unique")
    _validate_json(input_value.get("artifact"))
    return {"schema": POLICY_TEMPLATE_V1_SCHEMA, "template_id": _text(input_value.get("template_id"), "template_id", 256), "version": _text(input_value.get("version"), "version", 256), "parameters": parameters, "artifact": input_value["artifact"]}


def policy_template_v1_hash(input_value: Dict[str, Any]) -> str:
    return _digest("obsvr-policy-template/1", build_policy_template_v1(input_value))


def _param(parameter: Dict[str, Any], value: Any) -> Any:
    if parameter["type"] == "string" and isinstance(value, str):
        return _text(value, parameter["name"])
    if parameter["type"] == "integer" and isinstance(value, int) and not isinstance(value, bool) and abs(value) <= 9_007_199_254_740_991:
        return value
    if parameter["type"] == "boolean" and isinstance(value, bool):
        return value
    if parameter["type"] == "enum" and isinstance(value, str) and value in parameter["enum_values"]:
        return value
    _fail(f"parameter {parameter['name']} does not match type {parameter['type']}")


def _render(value: Any, params: Dict[str, Any]) -> Any:
    if isinstance(value, list):
        return [_render(item, params) for item in value]
    if isinstance(value, dict):
        if set(value) == {"$obsvr_param"}:
            name = value["$obsvr_param"]
            if not isinstance(name, str) or name not in params:
                _fail("template references an undeclared parameter")
            return params[name]
        return {key: _render(item, params) for key, item in value.items()}
    return value


def render_policy_template_v1(template_input: Dict[str, Any], supplied: Dict[str, Any], approval_hash: str, activation_hash: str) -> Dict[str, Any]:
    template = build_policy_template_v1(template_input)
    if not isinstance(supplied, dict):
        _fail("supplied parameters must be an object")
    expected = {item["name"] for item in template["parameters"]}
    extras = sorted(set(supplied) - expected, key=code_point_key)
    if extras:
        _fail(f"undeclared parameter supplied: {extras[0]}")
    params = {}
    for parameter in template["parameters"]:
        name = parameter["name"]
        if name not in supplied:
            _fail(f"missing parameter: {name}")
        params[name] = _param(parameter, supplied[name])
    params = dict(sorted(params.items(), key=lambda item: code_point_key(item[0])))
    rendered = _render(template["artifact"], params)
    _validate_json(rendered)
    return {"schema": RENDERED_POLICY_V1_SCHEMA, "template_id": template["template_id"], "template_version": template["version"], "template_hash": policy_template_v1_hash(template), "parameters": params, "parameters_hash": _digest("obsvr-policy-template-parameters/1", params), "rendered_artifact": rendered, "artifact_hash": _digest("obsvr-rendered-policy-artifact/1", rendered), "approval_hash": _hash(approval_hash, "approval_hash"), "activation_hash": _hash(activation_hash, "activation_hash")}


def sign_rendered_policy_v1(rendered: Dict[str, Any], signer: DeviceSigner) -> Dict[str, Any]:
    body_hash = _digest("obsvr-rendered-policy/1", rendered)
    payload = f"obsvr-rendered-policy-signature/1\0{body_hash}"
    return {"schema": RENDERED_POLICY_ENVELOPE_V1_SCHEMA, "body": rendered, "body_hash": body_hash, "key_id": signer.key_id, "signature": signer.sign_payload(payload)}


def verify_rendered_policy_v1(envelope: Dict[str, Any], raw_public_key: bytes) -> bool:
    try:
        body_hash = _digest("obsvr-rendered-policy/1", envelope["body"])
        return envelope.get("schema") == RENDERED_POLICY_ENVELOPE_V1_SCHEMA and envelope.get("body_hash") == body_hash and envelope.get("key_id") == derive_device_key_id(raw_public_key) and verify_device_sig(raw_public_key, envelope["key_id"], f"obsvr-rendered-policy-signature/1\0{body_hash}", envelope.get("signature")) is True
    except (KeyError, TypeError, ValueError):
        return False
